src/lecture_feedback/app.py

```diff
@@ -1,17 +1,76 @@
+import time
 import uuid
 
 import streamlit as st
 from streamlit_autorefresh import st_autorefresh
 
+from lecture_feedback.thread_safe_dict import ThreadSafeDict
 from lecture_feedback.user_stats_tracker import (
     UserStatsTracker,
     UserStatus,
 )
 
 
 @st.cache_resource
-def get_user_stats_tracker() -> UserStatsTracker:
-    return UserStatsTracker()
+def get_session_store() -> ThreadSafeDict:
+    """Return a shared map of shared_session_id -> UserStatsTracker.
+
+    This is stored as a cached resource so it's shared across reruns and
+    across users in the same Streamlit process.
+    """
+    return ThreadSafeDict()
+
+
+@st.cache_resource
+def get_cleanup_throttle() -> ThreadSafeDict:
+    """Store the last cleanup timestamp to throttle cleanup runs.
+
+    Returns a dict with key 'last_cleanup_time' holding a float timestamp.
+    """
+    return ThreadSafeDict({"last_cleanup_time": 0.0})
+
+
+def add_tracker_for_session(shared_session_id: str) -> None:
+    store = get_session_store()
+    if shared_session_id in store:
+        msg = f"Session ID '{shared_session_id}' already exists"
+        raise ValueError(msg)
```

Comment:
I think (next PR) this is not an exception its expected logic so we should thread it like that (no crash!)

Reply:
make sense